File: BOJ/17609_1.cpp
// 회문
// https://www.acmicpc.net/problem/17609
#include <iostream>
#include <algorithm>
#include <string>
using namespace std;
int N;
string str;
int alpha[26];
bool sudoPalin(int left, int right, bool flag)
{
    if (left > right)
        return true;
    if (str[left] == str[right])
        return sudoPalin(left + 1, right - 1, flag);
    else if (flag)
        return max(sudoPalin(left + 1, right, false), sudoPalin(left, right - 1, false));
    else
        return false;
}
bool palin(string str)
{
    int left = 0, right = str.size() - 1;
    while (left <= right)
    {
        if (str[left] != str[right])
            return false;
        left += 1;
        right -= 1;
    }
    return true;
}
int main()
{
    ios::sync_with_stdio(0);
    cin.tie(0);
    cout.tie(0);

    cin >> N;
    for (int i = 0; i < N; i++)
    {
        cin >> str;
        // 회문인 경우
        if (palin(str))
            cout << 0 << '\n';
        // 유사 회문 체크
        else
        {
            // str크기가 홀수인 경우 =>  유사 회문인 경우
            if (sudoPalin(0, str.size() - 1, true))
                cout << 1 << '\n';
            else
                cout << 2 << '\n';
        }
    }
    return 0;
}

/*
    참고
    https://newdeal123.tistory.com/13
*/

File: BOJ/17609_1.py
'''
    회문
    https://www.acmicpc.net/problem/17609
'''
import sys
for _ in range(int(sys.stdin.readline())):
    s = sys.stdin.readline().strip()
    # 문자열 앞과, 문자열 뒤를 한개씩 비교
    if s == s[::-1]:
        print(0)
    else:
        index = 0
        # 일치하지 않는 곳의 index를 찾는다.
        while s[index] == s[-index-1]:
            index += 1
        # 앞으로 한 칸 이동한 문자열
        u = s[index + 1: len(s) - index]
        # 뒤에서 한 칸 이동한 문자열
        v = s[index: -index - 1]
        print(1 if u == u[::-1] or v == v[::-1] else 2)
